test/MediatR.Tests/GenericTypeConstraintsTests.cs

```diff
@@ -0,0 +1,157 @@
+            {
+                cfg.Scan(scanner =>
+                {
+                    scanner.AssemblyContainingType(typeof(GenericTypeConstraintsTests));
+                    scanner.IncludeNamespaceContainingType<Ping>();
+                    scanner.IncludeNamespaceContainingType<Jing>();
+                    scanner.WithDefaultConventions();
+                    scanner.AddAllTypesOf(typeof(IRequestHandler<,>));
+                    scanner.AddAllTypesOf(typeof(IRequestHandler<>));
+                });
+                cfg.For<SingleInstanceFactory>().Use<SingleInstanceFactory>(ctx => t => ctx.GetInstance(t));
+                cfg.For<MultiInstanceFactory>().Use<MultiInstanceFactory>(ctx => t => ctx.GetAllInstances(t));
+                cfg.For<IMediator>().Use<Mediator>();
+            });
+
+            _mediator = container.GetInstance<IMediator>();
+        }
+
+        [Fact]
+        public async Task Should_Resolve_Void_Return_Request()
+        {
+            // Create Request
+            var jing = new Jing { Message = "Jing" };
+
+            // Test mediator still works sending request
+            await _mediator.Send(jing);
+
+            // Create new instance of type constrained class
+            var genericTypeConstraintsVoidReturn = new  GenericTypeConstraintJing();
+
+            // Assert it is of type IRequest and IRequest<T>
+            Assert.True(genericTypeConstraintsVoidReturn.IsIRequest);
+            Assert.True(genericTypeConstraintsVoidReturn.IsIRequestT);
+            Assert.True(genericTypeConstraintsVoidReturn.IsIBaseRequest);
+
+            // Verify it is of IRequest
+            genericTypeConstraintsVoidReturn.Handle(jing)
+                .Select(x => x.ShouldBeOfType<IRequest>())
+                .Select(x => x.ShouldBeOfType<IRequest<Unit>>())
+                .Select(x => x.ShouldBeOfType<IBaseRequest>());
+        }
+
+        [Fact]
+        public async Task Should_Resolve_Response_Return_Request()
+        {
+            // Create Request
+            var ping = new Ping { Message = "Ping" };
+
+            // Test mediator still works sending request and gets response
+            var pingResponse = await _mediator.Send(ping);
+            pingResponse.Message.ShouldBe("Ping Pong");
+
+            // Create new instance of type constrained class
+            var genericTypeConstraintsResponseReturn = new GenericTypeConstraintPing();
+
+            // Assert it is of type IRequest<T> but not IRequest
+            Assert.False(genericTypeConstraintsResponseReturn.IsIRequest);
+            Assert.True(genericTypeConstraintsResponseReturn.IsIRequestT);
+            Assert.True(genericTypeConstraintsResponseReturn.IsIBaseRequest);
+
+            // Verify it is of IRequest<Pong>
+            genericTypeConstraintsResponseReturn.Handle(ping)
+                .Select(x => x.ShouldBeOfType<IRequest<Pong>>())
+                .Select(x => x.ShouldBeOfType<IRequest>())
```

Comment:
this should not be IRequest right?

Reply:
Sorry this test was actually broken, I've updated it to actually test the scenario correctly now.
But now ping should not be be of `IRequest`, it should be if of `IRequest<Pong>` and `IBaseRequest`

the updated version now accurately shows this.